Stop special casing LogAlways

noahfalk opened the pull request:
Fixes #51429

LogAlways is the name for the most severe logging level but the code was mistakenly treating it as a special case that skipped
all level filtering. I eliminated that special case handling.


@brianrob:
 - does this match your understanding of how a listener at LogAlways filtering level is intended to behave?
 - any concerns that we should avoid fixing this because of the back-compat implications?

@josalem - could you review this?
cc @dotnet/dotnet-diag

noahfalk committed:
Stop special casing LogAlways

Fixes #51429

LogAlways is the name for the most severe logging level but
the code was mistakenly treating it as a special case that skipped
all level filtering. I eliminated that special case handling.

josalem commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:
```diff
@@ -2308,7 +2308,7 @@ private bool IsEnabledByDefault(int eventNum, bool enable, EventLevel currentLev
                 return false;
 
             // does is pass the level test?
```
```suggestion
            // does it pass the level test?
```

noahfalk replied:
fixed : )

josalem reviewed (approved):
This matches with my understanding of what LogAlways is supposed to mean. We should make sure our examples and docs all reflect this.

brianrob commented:
Yes, this matches my understanding as well.

noahfalk committed:
Fix comment typo

noahfalk committed:
Fixing EventSource test issues

- One test had a listener that dependended on LogAlways to mean "don't filter". Changed it to use Verbose instead.
- Cleaned up a few other test issues that were obfuscating my local test run with spurious failures

stephentoub commented on src/libraries/System.Diagnostics.Tracing/tests/BasicEventSourceTest/LoudListener.cs:
```diff
@@ -18,7 +18,7 @@ internal class LoudListener : EventListener
 
         public LoudListener(EventSource eventSource)
         {
-            EnableEvents(eventSource, EventLevel.LogAlways, (EventKeywords)0xffffffff);
```
I'm confused.  I thought `EnableEvents(eventSource, EventLevel.LogAlways, (EventKeywords)0xffffffff);` was basically saying "enable everything"... is it not?

stephentoub replied:
If that's not the case, or if this is making that not be the case, I fear the fallout will be huge.  Just in dotnet/runtime, don't all of these assume the same thing?
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Diagnostics.DiagnosticSource/tests/DiagnosticSourceEventSourceBridgeTests.cs#L1525-L1529
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/Common/tests/System/Diagnostics/Tracing/ConsoleEventListener.cs#L22
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Net.Http/tests/StressTests/HttpStress/HttpEventListener.cs#L43
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/XplatEventLogger.cs#L167

noahfalk replied:
There is a bunch of confusion around this, and I am trying to get a better picture of the implications. For the moment the change is on hold.

Different people think it means different things. The rule that every other level follows is that events get logged when level_of_event <= level_of_filter. Since LogAlways = 0, that implies it is the most restrictive filter. Our docs also agree with that interpretation. However on the flip side we've got code examples like the ones you found that assume LogAlways is a special case value that is the most permissive filter and certainly other code in the wild could make the same assumptions. I am chatting on email with the ETW folks because there is also evidence they treat the filter with LogAlways as a special case.

Another possible outcome is that we go change all our docs to say that LogAlways is a special case. I'm trying to hunt down other places in code we've made this check to see if they consistently special case it though I am fearful we are probably applying inconsistent rules in different code paths.

noahfalk commented:
I concluded we are better off leaving the existing behavior intact and fixing the documentation to match it rather than changing the behavior.
